app/src/main/java/org/mozilla/fenix/perf/ProfilerUtils.kt

```diff
@@ -0,0 +1,147 @@
+    Networking(networking_threads, networking_features);
+}
+
+/**
+ * Collection of functionality to parse and save the profile returned by GeckoView.
+ */
+object ProfilerUtils {
+
+    private fun saveProfileUrlToClipboardAndToast(profileResult: ByteArray, context: Context): String {
+        // The profile is saved to a temporary file since our fetch API takes a file or a string.
+        // Converting the ByteArray to a String would hurt the encoding, which we need to preserve.
+        val outputFile = createTemporaryFile(profileResult, context)
+        val response = networkCallToProfilerServer(outputFile, context)
+        val profileToken = decodeProfileToken(response)
+        outputFile.delete()
+        return PROFILER_DATA_URL + profileToken
+    }
+
+    private fun finishProfileSave(context: Context, url: String, onUrlFinish: (Int) -> Unit) {
+        val clipboardManager = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
+        val clipData = ClipData.newPlainText("Profiler URL", url)
+        clipboardManager.setPrimaryClip(clipData)
+        onUrlFinish(R.string.profiler_uploaded_url_to_clipboard)
+    }
+
+    private fun createTemporaryFile(profileResult: ByteArray, context: Context): File {
+        val outputDir = context.cacheDir
+        val outputFile = File.createTempFile("tempProfile", ".json", outputDir)
+        FileOutputStream(outputFile).use { fileOutputStream ->
+            fileOutputStream.write(profileResult)
+            return outputFile
+        }
+    }
+
+    private fun networkCallToProfilerServer(outputFile: File, context: Context): Response {
+        val request = Request(
+            url = PROFILER_API,
+            method = Request.Method.POST,
+            headers = MutableHeaders(
+                "Accept" to PROFILER_SERVER_HEADER
+            ),
+            body = Request.Body.fromFile(outputFile)
+        )
+        return context.components.core.client.fetch(request)
+    }
+
+    private fun decodeProfileToken(response: Response): String {
+        val jwtToken = StringBuilder()
+        response.body.useBufferedReader {
+            val jwt = it.readText()
+            val jwtSplit = jwt.split(".")
+            val decodedBytes = Base64.decode(jwtSplit[1], Base64.DEFAULT)
+            val decodedStr = decodedBytes.decodeToString()
+            val jsonObject = JSONObject(decodedStr)
+            jwtToken.append(jsonObject[TOKEN])
+        }
+        return jwtToken.toString()
+    }
+
+    /**
+     * This will either save the profile locally or send it as a URL to the Firefox profiler server
+     *
+     * @param context Activity context to get access to the profiler API through components.core...
+     * @param profile Data returned from GeckoView as a GZIP ByteArray
+     * @param onUrlFinish function passed in to display a toast with the relevant information once the profile is saved
+     */
+    fun handleProfileSave(
```

Comment:
thought: it feels kind of scary to me how easy it is to upload sensitive data with the profiler. Maybe we should check with Bas or someone else on the performance team that the UX of this change is acceptable.

Reply:
As discussed, we should also consider translating the strings so users outside EN locales understand what is happening.